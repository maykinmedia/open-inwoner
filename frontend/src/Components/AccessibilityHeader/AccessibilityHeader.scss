.accessibility-header {
  $hm: max(calc((100vw - var(--container-width)) / 2), var(--spacing-large));
  background-color: var(--color-accent);
  display: grid;
  grid-template-areas: "ml-1 lst mr-1";
  grid-template-columns: $hm 1fr $hm;

  /// Layout.

  &__list {
    grid-area: lst;
  }

  /// List.

  &__list {
    display: flex;
    list-style: none;
    margin: 0;
    padding: 0;
  }

  &__list-item {
    margin: 0 var(--spacing-large);

    &:first-child {
      margin-left: 0;
    }

    &:last-child {
      margin-right: 0;
    }
  }

  /// Link.

  .link {
    line-height: var(--row-height);
    @media (max-width: 767px) {

      &:not(:focus):not(:hover) {
        font-size: 0;
      }
    }
  }
}
