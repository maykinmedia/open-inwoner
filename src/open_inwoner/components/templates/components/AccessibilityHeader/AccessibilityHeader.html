{% load i18n components %}

<header class="accessibility-header">
    <ul class="accessibility-header__list">
        <li class="accessibility-header__list-item">
            {% link href="#" icon="hearing" %}{% trans "Lees voor" %}{% endlink %}
        </li>

        <li class="accessibility-header__list-item">
            {% link href="#" icon="zoom_in" %}
                {% trans "Vergroten" %}
            {% endlink %}
        </li>

        <li class="accessibility-header__list-item">
            {% link href="#" icon="font_download" %}{% trans "Dyslexie" %}{% endlink %}
        </li>

        <li class="accessibility-header__list-item">
            {% link href="#" icon="print" %}{% trans "Print pagina" %}{% endlink %}
        </li>
    </ul>
</header>
