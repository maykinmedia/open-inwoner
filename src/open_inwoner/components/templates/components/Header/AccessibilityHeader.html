{% load i18n link_tags %}

<header class="accessibility-header">
    <ul class="accessibility-header__list">
        <li class="accessibility-header__list-item">
            {% trans "Lees voor" as link_text %}
            {% link href="#" icon="hearing" text=link_text extra_classes="accessibility--read" %}
        </li>

        <li class="accessibility-header__list-item">
            {% trans "Vergroten" as link_text %}
            {% trans "Verkleinen" as alt_link_text %}
            {% link href="#" icon="zoom_in" text=link_text extra_classes="accessibility--enlarge-font" data_text=link_text data_alt_text=alt_link_text data_icon="zoom_in" data_alt_icon="zoom_out" %}
        </li>

        <li class="accessibility-header__list-item">
            {% trans "Dyslexie" as link_text %}
            {% trans "Normaal lettertype" as alt_link_text %}
            {% link href="#" icon="font_download" text=link_text extra_classes="accessibility--change-font" data_text=link_text data_alt_text=alt_link_text %}
        </li>

        <li class="accessibility-header__list-item">
            {% trans "Print pagina" as link_text %}
            {% link href="#" icon="print" text=link_text extra_classes="accessibility--print" %}
        </li>
    </ul>
</header>
