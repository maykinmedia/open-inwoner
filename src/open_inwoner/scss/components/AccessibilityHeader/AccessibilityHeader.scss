.accessibility-header {
  $hm: max(calc((100vw - var(--container-width)) / 2), var(--spacing-large));
  background-color: var(--color-accent);
  box-sizing: border-box;
  padding: 0 $hm;

  /// List.

  &__list {
    display: flex;
    list-style: none;
    margin: 0;
    padding: 0;
  }

  &__list-item {
    margin: 0 var(--spacing-large);

    &:first-child {
      margin-left: 0;
    }

    &:last-child {
      margin-right: 0;
    }
  }

  /// Link.

  .link {
    color: var(--color-font-accent);
    line-height: var(--row-height);
    @media (max-width: 767px) {

      &:not(:focus):not(:hover) {
        font-size: 0;
      }
    }
  }
}
