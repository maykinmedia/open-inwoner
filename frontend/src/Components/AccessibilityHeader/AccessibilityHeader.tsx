import React, {ReactElement, useRef, useState} from 'react';
import FontDownloadOutlinedIcon from '@mui/icons-material/FontDownloadOutlined';
import HearingOutlinedIcon from '@mui/icons-material/HearingOutlined';
import PrintOutlinedIcon from '@mui/icons-material/PrintOutlined';
import ZoomInOutlinedIcon from '@mui/icons-material/ZoomInOutlined';
import ZoomOutOutlinedIcon from '@mui/icons-material/ZoomOutOutlined';
import {Link} from '../Typography/Link';
import './AccessibilityHeader.scss';


/**
 * Header containing various accessibility features.
 * @return {ReactElement}
 */
export function AccessibilityHeader(): ReactElement {
  const speechSynthesis = (typeof window === 'object') ? window['speechSynthesis'] : null;
  const SpeechSynthesisUtterance = (typeof window === 'object') ? window['SpeechSynthesisUtterance'] : null;
  const isSpeechSynthesisSupported = Boolean(speechSynthesis && SpeechSynthesisUtterance);
  const utteranceRef = useRef<{ current: SpeechSynthesisUtterance | null }>(null);
  const [previousFontSize, setPreviousFontSize] = useState<string | null>(null);
  const [previousFontFamily, setPreviousFontFamily] = useState<string | null>(null);

  /**
   * Runs text to speech ont the textContent of <main>.
   */
  const textToSpeech = (): void => {
    if (!isSpeechSynthesisSupported) {
      return
    }

    const main = (typeof document === 'undefined') ? null : document.querySelector('main');
    const mainTextContent = main?.textContent;

    if (!utteranceRef.current) {
      // @ts-ignore
      utteranceRef.current = new SpeechSynthesisUtterance('');
    }
    const utterance = utteranceRef?.current as any;
    utterance.text = mainTextContent;
    speechSynthesis?.speak(utterance);
  }

  /**
   * Toggles zoom of regular text.
   */
  const zoom = (): void => {
    const target = document.querySelector(':root') as HTMLElement;
    const varName = '--font-size-body';

    if (!target) {
      return;
    }

    if (previousFontSize) {
      target.style.removeProperty(varName);
      setPreviousFontSize(null);
      return;
    }

    const currentFontSize = getComputedStyle(target).getPropertyValue(varName).trim()
    const currentFontSizeValue = parseInt(currentFontSize);
    const currentFontSizeUnitMatch = currentFontSize.match(/[^\d]+/);
    const currentFontSizeUnit = (currentFontSizeUnitMatch) ? currentFontSizeUnitMatch[0] : '';
    const newFontSizeValue = currentFontSizeValue * 1.2;
    const newFontSize = `${newFontSizeValue}${currentFontSizeUnit}`;

    setPreviousFontSize(currentFontSize);
    target?.style.setProperty(varName, newFontSize);
  }

  /**
   * Swaps the font for a set of font more optimized for people with Dyslexia.
   */
  const swapFont = (): void => {
    const target = document.querySelector(':root') as HTMLElement;
    const varName = '--font-family-body';
    const alternativeFontFamily = `Helvetica, Courier, Arial, Verdana.`;

    if (!target) {
      return;
    }

    if (previousFontFamily) {
      target.style.removeProperty(varName);
      setPreviousFontFamily(null);
      return;
    }

    const currentFontFamily = getComputedStyle(target).getPropertyValue(varName).trim()

    setPreviousFontFamily(currentFontFamily);
    target?.style.setProperty(varName, alternativeFontFamily);
  }

  return (
    <header className="accessibility-header">
      <ul className="accessibility-header__list">
        {isSpeechSynthesisSupported && <li className="accessibility-header__list-item">
          <Link to="#" icon={HearingOutlinedIcon} onClick={textToSpeech}>Lees voor</Link>
        </li>}

        <li className="accessibility-header__list-item">
          <Link to="#" icon={(previousFontSize) ? ZoomOutOutlinedIcon : ZoomInOutlinedIcon} onClick={zoom}>
            Vergroten
          </Link>
        </li>

        <li className="accessibility-header__list-item">
          <Link to="#" icon={FontDownloadOutlinedIcon} onClick={swapFont}>
            {(previousFontFamily) ? 'Normaal lettertype' : 'Dyslexie'}
          </Link>
        </li>

        <li className="accessibility-header__list-item">
          <Link to="javascript:print()" icon={PrintOutlinedIcon}>Print pagina</Link>
        </li>
      </ul>
    </header>
  )
}
