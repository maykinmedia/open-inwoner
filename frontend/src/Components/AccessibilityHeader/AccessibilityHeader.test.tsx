import React from 'react';
import renderer from 'react-test-renderer';
import { AccessibilityHeader } from './AccessibilityHeader';

test('Render accessibility header', () => {
  const component = renderer.create(
    <AccessibilityHeader />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Click zoom accessibility header', () => {
  const component = renderer.create(
    <AccessibilityHeader />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  console.log(tree.children[0].children[0].children[0]);
  tree.children[0].children[0].children[0].onClick();
  expect(tree).toMatchSnapshot();
});
