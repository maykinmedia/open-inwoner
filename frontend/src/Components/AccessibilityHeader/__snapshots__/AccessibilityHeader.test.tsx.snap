// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render accessibility header 1`] = `
<header
  className="accessibility-header"
>
  <ul
    className="accessibility-header__list"
  >
    <li
      className="accessibility-header__list-item"
    >
      <a
        className="link"
        href="#"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ZoomInOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14zm.5-7H9v2H7v1h2v2h1v-2h2V9h-2z"
          />
        </svg>
        Vergroten
      </a>
    </li>
    <li
      className="accessibility-header__list-item"
    >
      <a
        className="link"
        href="#"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="FontDownloadOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9.17 15.5h5.64l1.14 3h2.09l-5.11-13h-1.86l-5.11 13h2.09l1.12-3zM12 7.98l2.07 5.52H9.93L12 7.98zM20 2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 18H4V4h16v16z"
          />
        </svg>
        Dyslexie
      </a>
    </li>
    <li
      className="accessibility-header__list-item"
    >
      <a
        className="link"
        href="#"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="PrintOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 8h-1V3H6v5H5c-1.66 0-3 1.34-3 3v6h4v4h12v-4h4v-6c0-1.66-1.34-3-3-3zM8 5h8v3H8V5zm8 12v2H8v-4h8v2zm2-2v-2H6v2H4v-4c0-.55.45-1 1-1h14c.55 0 1 .45 1 1v4h-2z"
          />
          <circle
            cx="18"
            cy="11.5"
            r="1"
          />
        </svg>
        Print pagina
      </a>
    </li>
  </ul>
</header>
`;
